website/project/views/node.py

```diff
@@ -833,6 +844,107 @@ def _get_children(node, auth, indent=0):
     return children
 
 
+def n_unread_comments(node, user, page, root_id=None, check=False):
+    """Return the number of unread comments on a node for a user."""
+    if not node.is_contributor(user):
+        return 0
+    if root_id is None or page == 'node':
+        return n_unread_total(node, user, page, check=check)
+    root_target = Guid.load(root_id)
+    if root_target:
+        root_target = root_target.referent
+    else:
+        root_target = node.get_wiki_page(root_id, 1)
+    if page == 'files' and check:
+        exists, _ = check_file_exists(node, root_id)
+        if not exists:
+            return 0
+    default_timestamp = datetime(1970, 1, 1, 12, 0, 0)
+    view_timestamp = user.comments_viewed_timestamp.get(node._id, default_timestamp)
+    if isinstance(view_timestamp, dict):
+        view_timestamp = view_timestamp.get(page, default_timestamp)
+    if not page == 'node' and isinstance(view_timestamp, dict):
```

Comment:
This conditional logic to handle different page types is repeated in multiple places. Could this be refactored into `Node` methods?

```python
node.get_user_comment_timestamp(user, page, root_id=None...)
node.set_user_comment_timestamp(user, page, root_id=None...)
```